migration/capcons/migration_test.go

```diff
@@ -0,0 +1,1556 @@
+				BorrowType: testRReferenceStaticType,
+				Path: interpreter.PathValue{
+					Domain:     common.PathDomainPrivate,
+					Identifier: testPathIdentifier,
+				},
+				Address: interpreter.AddressValue(testAddress),
+			},
+			pathLinks: []testLink{
+				{
+					// Equivalent to:
+					//   link<&Test.R>(/private/test, target: /storage/test)
+					sourcePath: interpreter.PathValue{
+						Domain:     common.PathDomainPrivate,
+						Identifier: testPathIdentifier,
+					},
+					targetPath: interpreter.PathValue{
+						Domain:     common.PathDomainStorage,
+						Identifier: testPathIdentifier,
+					},
+					borrowType: testRReferenceStaticType,
+				},
+			},
+			expectedMigrations: []testCapConsPathCapabilityMigration{
+				{
+					accountAddress: testAddress,
+					addressPath: interpreter.AddressPath{
+						Address: testAddress,
+						Path: interpreter.NewUnmeteredPathValue(
+							common.PathDomainPrivate,
+							testPathIdentifier,
+						),
+					},
+					borrowType: testRReferenceStaticType,
+				},
+			},
+		},
+		// Test that the migration also follows capability controller,
+		// which were already previously migrated from links.
+		// Following the (capability value) should not borrow it,
+		// i.e. require the storage target to exist,
+		// but rather just get the storage target
+		{
+			name: "Path links, working chain (private -> private -> storage)",
+			// Equivalent to: getCapability<&Test.R>(/private/test)
+			capabilityValue: &interpreter.PathCapabilityValue{ //nolint:staticcheck
+				BorrowType: testRReferenceStaticType,
+				Path: interpreter.PathValue{
+					Domain:     common.PathDomainPrivate,
+					Identifier: testPathIdentifier,
+				},
+				Address: interpreter.AddressValue(testAddress),
+			},
+			pathLinks: []testLink{
+				{
+					// Equivalent to:
+					//   link<&Test.R>(/private/test, target: /private/test2)
+					sourcePath: interpreter.PathValue{
+						Domain:     common.PathDomainPrivate,
+						Identifier: testPathIdentifier,
+					},
+					targetPath: interpreter.PathValue{
+						Domain:     common.PathDomainPrivate,
+						Identifier: "test2",
+					},
+					borrowType: testRReferenceStaticType,
+				},
+				{
+					// Equivalent to:
+					//   link<&Test.R>(/private/test2, target: /storage/test)
+					sourcePath: interpreter.PathValue{
+						Domain:     common.PathDomainPrivate,
+						Identifier: "test2",
+					},
+					targetPath: interpreter.PathValue{
+						Domain:     common.PathDomainStorage,
+						Identifier: testPathIdentifier,
+					},
+					borrowType: testRReferenceStaticType,
+				},
+			},
+			expectedMigrations: []testCapConsPathCapabilityMigration{
+				{
+					accountAddress: testAddress,
+					addressPath: interpreter.AddressPath{
+						Address: testAddress,
+						Path: interpreter.NewUnmeteredPathValue(
+							common.PathDomainPrivate,
+							testPathIdentifier,
+						),
+					},
+					borrowType: testRReferenceStaticType,
+				},
+			},
+		},
+		// TODO: verify
```

Comment:
This test case should fail. When migrating, perform a "borrow" (i.e. subtype check) against the Cap Con.
Also, add a test case for borrowing with e.g. an restricted type (interface), both matching and non-matching. Thank you for suggesting that, @SupunS!